Add message sending function

tokuchan-pg opened the pull request:
# WHAT
メッセージ送信機能の実装
- メッセージ本文が未入力の場合はテーブルに保存されないようにバリデーションを設定
- messages#createアクションをバリデーションの結果に基づいて成功時と失敗時に分岐させ、それに応じて適切なフラッシュメッセージと画面遷移を設定

# WHY
チャットアプリに必要な機能のため

## 補足
同時に、以下の変更も加えました。
- トップページのパスをmessages#indexからgroups#indexに変更（左カラムにログインユーザーが参加しているグループ一覧を表示させ、メッセージ表示部分には何も表示させない）
- 部分テンプレートによるビューのリファクタリング


### メッセージ送信成功時↓
![2017-04-26 11 58 13](https://cloud.githubusercontent.com/assets/25572309/25416231/b55e8216-2a77-11e7-803d-03cf6a143c57.png)


### メッセージ送信失敗時↓
![2017-04-26 11 58 22](https://cloud.githubusercontent.com/assets/25572309/25416247/c7fccfcc-2a77-11e7-8c55-8fef9ad74cf3.png)


### トップページ（groups#index）↓
![2017-04-26 11 23 50](https://cloud.githubusercontent.com/assets/25572309/25416155/3de46412-2a77-11e7-997d-7ec53044a9e2.png)


### チャット画面（messages#index）↓
![2017-04-26 11 19 59](https://cloud.githubusercontent.com/assets/25572309/25416187/67e62caa-2a77-11e7-8c69-39ee59322410.png)


### ビューファイルと部分テンプレートの構成状況↓
- ↓の画面キャプチャーにて、開いている４つのビューファイルのディレクトリをハイライトさせています
- groups#indexとmessages#indexで共通で使用する部分テンプレートはsharedディレクトリにまとめました

**（１、左上）**
views/groups/index.html/haml
※トップページ　→　下記３を読み込んでいる

**（２、右上）**
views/messages/index.html/haml
※特定のグループのチャット画面　→　下記３を読み込んでいる

**（３、左下）**
views/**shared**/_chat-side.html/haml
※左カラム全体を表示する部分テンプレート　→　下記４を読み込んでいる

**（４、右下）**
views/**shared**/_group.html/haml
※左カラムの中の、ログインユーザーのグループ一覧を表示する部分テンプレート

![2017-04-26 11 18 21](https://cloud.githubusercontent.com/assets/25572309/25416293/0899ac44-2a78-11e7-8f65-4d625e8271ba.png)

tokuchan-pg committed:
generate message model and messages table

tokuchan-pg committed:
defined association

tokuchan-pg committed:
defined routing for messages controller

tokuchan-pg committed:
changed the root page from messages#index to groups#index

At the same time,  below two updates in groups_controller.
*Defined index action
*Updated the redirect path of create and update action

tokuchan-pg committed:
generated messages_controller and defined index & create action

At the same time, defined validation in message model. Create action
shows appropriate flash message after a message saved or not.

tokuchan-pg committed:
updated the views by using partial templates

Updated  below two views
*groups/index (new root page from this branch)
*messages/index (former root page)

motemotenten commented on app/controllers/groups_controller.rb:
```diff
@@ -2,14 +2,17 @@ class GroupsController < ApplicationController
 
   before_action :set_group, only: [:edit, :update]
 
+  def index
```
こちらこういった書き方もできます。
```
def index; end
```

tokuchan-pg replied:
修正いたしました。

motemotenten commented on app/controllers/groups_controller.rb:
```diff
@@ -21,7 +24,7 @@ def edit
   def update
     if @group.update(group_params)
-      redirect_to root_path, notice: 'グループが更新されました'
+      redirect_to group_messages_path(@group), notice: 'グループが更新されました'
```
こちら補足なのですが、rails5からは、 インスタンスを引数に取らなくても動きます。こちらの方が明示的なので、わかりやすいですが。

tokuchan-pg replied:
今回は自分の理解のために明示した状態で進めさせて頂けたらと思います。今後は省略いたします！

motemotenten commented on app/controllers/messages_controller.rb:
```diff
@@ -1,4 +1,33 @@
+  end
+
+  def set_messages
+    @messages = Message.where(group_id: params[:group_id])
```
アソシエーションを使って取ってきましょう。

tokuchan-pg replied:
修正いたしました。includesを使ってn+1問題にも対応させました。

motemotenten commented on app/controllers/messages_controller.rb:
```diff
@@ -1,4 +1,33 @@
   end
+
+  def create
+    @message = Message.new(message_params)
```
current_userのアソシエーションを使うことでインスタンスを生成しましょう

tokuchan-pg replied:
修正いたしました。

motemotenten commented on app/controllers/messages_controller.rb:
```diff
@@ -1,4 +1,33 @@
 class MessagesController < ApplicationController
+
+  before_action :set_group
+  before_action :set_messages
```
onlyなどで、メソッドの適用を限定しましょう。

tokuchan-pg replied:
修正致しました。

motemotenten commented on app/models/group.rb:
```diff
@@ -5,6 +5,8 @@ class Group < ApplicationRecord
+  has_many :messages
+
   # グループを新規作成・更新したときに、アソシエーション先であるusersも含めて保存・更新できるようにする
   accepts_nested_attributes_for :users
```
こちら記述しなくても、動きます。理由としては、グループを新規作成・更新したときに、userを新規作成するわけではないからです。
http://qiita.com/hmuronaka/items/818c421dc632e3efb7a6

tokuchan-pg replied:
修正いたしました。

tokuchan-pg committed:
made the code with no processing more simple

tokuchan-pg committed:
used association to set instance variable (with 'incluedes' to solve n+1 issue)

tokuchan-pg committed:
used association to set instance variable

tokuchan-pg committed:
used 'only' option for before_action

tokuchan-pg committed:
removed accepts_nested_attributes_for

When creating or updating a group, it doesn’t mean creating new user,
so accepts_nested_attributes_for is unneeded.
